File: 遗传优化/差分进化算法/DE.py
from collections import namedtuple
import math as Math
import random
from typing import Iterable, List
import numpy as np


# indi = namedtuple("indi", ["x", "fitness"])
class indi(object):
    def __init__(self, x:Iterable, fitness:float):
        self.x = x
        self.fitness = fitness
    
    def clone(self):
        return indi(self.x, self.fitness)

def six_hump_camel_back(x):
    return 4*Math.pow(x[0], 2) - 2.1*Math.pow(x[0], 4)+ Math.pow(x[0],6)/3.0 + x[0]*x[1] - 4*Math.pow(x[1], 2) + 4*Math.pow(x[1], 4)

class DE(object):
    D = 2
    F = 0.5
    CR = 0.9
    LOW = [-3, -2]
    UP = [3, 2]
    PS = 30
    MaxFEs = 5E3
    
    # rnd, FEs, pop, bestIdx
    
    def __init__(self):
        # rnd = random
        # pop = [indi([0, 0], 0) for _ in range(self.PS)]
        self.FEs = 0
        self.pop: List[indi] = []
    
    def increaseFEs(self, count):
        self.FEs += count
    
    def _sample(self, count, idx):
        # t = 0
        # r = []
        # while t<count:
        #     num = random.randrange(0, self.PS)
        #     if num not in r and num!=idx:
        #         r.append(num)
        #         t += 1
        r = []
        l = [i for i in range(self.PS)]
        l.remove(idx)
        for i in range(count):
            tempidx = int(random.random() * len(l))
            r.append(l.pop(tempidx))
        return r
    
    def init(self):
        for _ in range(self.PS):
            vector = np.random.rand(2)
            self.pop.append(indi(vector, six_hump_camel_back(vector)))
        self.increaseFEs(self.PS)
    
    def iteration(self):
        while self.FEs <= self.MaxFEs:
            for i in range(self.PS):
                # trialindi = indi([0 for _ in range(self.D)], 0)
                # self.clone(trialindi, self.pop[i])
                trialindi = self.pop[i].clone()
                rndD = int(random.random())
                r = self._sample(3, i)
                for j in range(self.D):
                    num = random.random()
                    if num<=self.CR or rndD==j:
                        trialindi.x[j] = self.pop[r[0]].x[j] + self.F*(self.pop[r[1]].x[j]-self.pop[r[2]].x[j])
                        if trialindi.x[j]<self.LOW[j] or self.UP[j]<trialindi.x[j]:
                            trialindi.x[j] = self.LOW[j] + random.random()*(self.UP[j]-self.LOW[j])
                trialindi.fitness = six_hump_camel_back(trialindi.x)
                if trialindi.fitness < self.pop[i].fitness:
                    # self.clone(self.pop[i], trialindi)
                    self.pop[i] = trialindi.clone()
                self.increaseFEs(self.PS)
            
    def getBest(self):
        self.bestIdx = 0
        for i in range(self.PS):
            if self.pop[i].fitness < self.pop[self.bestIdx].fitness:
                self.bestIdx = i
        return self.pop[self.bestIdx]

sample = DE()
sample.init()
sample.iteration()
best = sample.getBest()
print(f"Best fitness={best.fitness}")
for i in range(sample.D):
    print(f"x_{i+1}={best.x[i]}", end=' ')
    if (i+1) % 5 == 0:
        print()
print(six_hump_camel_back(best.x))
# import matplotlib.pyplot as plt
# x = []
# y = []
# for i in range(100):
#     sample = DE()
#     sample.init()
#     sample.iteration()
#     best = sample.getBest()
#     x.append(best.x[0])
#     y.append(best.x[1])
# plt.scatter(x, y)
# plt.show()

File: 遗传优化/差分进化算法/DE/DE.java
/*
 * A demo of basic DE implementation.
 * 
 */

import java.util.ArrayList;
import java.util.Random;

public class DE {

	public static final int D = 2;  //dimensionality of benchmark function

	final double F = 0.5; //scaling factor
	final double CR = 0.9; //crossover probability
	final double[] LOW = { -3, -2 }; //upper boundary
	final double[] UP = { 3, 2 }; //lower boundary
	final int PS = 30; //population size
	final double MaxFEs = 5E3; //maximal number of function evaluations

	Random rnd; //generate a random number in [0, 1]
	int FEs; //used number of function evaluations
	indi[] pop; //population	
	int bestIdx; //index of the best individual


	public DE() {
		rnd = new Random();
		pop = new indi[PS];
	}

	public static void main(String[] args) {
		DE sample = new DE();
		sample.init();
		sample.iteration();
		sample.getBest();		
		System.out.printf("Best fitness=%f \n", sample.pop[sample.bestIdx].fitness);
		for (int i = 0; i < D; i++) {
			System.out.printf("x%d=%f   ", (i+1), sample.pop[sample.bestIdx].x[i]);
			if ((i + 1) % 5 == 0)
				System.out.println();
		}
	}

	
	/*
	 * To increase the number of function evaluations
	 * @count, how many times to increase
	 */
	public void increaseFEs(int count) {
		FEs += count;
	}



	/*
	 * Copy the @src individual to the @dst individual  
	 */
	public void clone(indi dst, indi src) {		
		dst.fitness = src.fitness;
		for (int i = 0; i < D; i++) {
			dst.x[i] = src.x[i];
		}
	}
	
	
	/*
	 * Save the indices of @count individuals into array @r, which are different from @idx.
	 */
	public void rndInt(int idx, int count, int[] r) {
		ArrayList<Integer> list = new ArrayList<Integer>();
		for (int i = 0; i < PS; i++) {
			list.add(i);
		}
		list.remove(idx);
		for (int i = 0; i < count; i++) {
			int tempidx = (int) Math.floor(rnd.nextDouble() * list.size());
			r[i] = list.get(tempidx);
			list.remove(tempidx);
		}
	}

	
	/*
	 * Initialization
	 */
	public void init() {
		FEs = 0;
		for (int i = 0; i < PS; i++) {
			pop[i] = new indi();
			for (int j = 0; j < D; j++) {
				pop[i].x[j] = LOW[j] + (UP[j]-LOW[j])*rnd.nextDouble();
			}
			pop[i].fitness = Six_hump_Camel_back.f(pop[i].x); //calculate the fitness value
		}
		increaseFEs(PS);
	}

	
	/*
	 * Iteration process of DE
	 */
	public void iteration() {
		while (FEs <= MaxFEs) {
			for (int i = 0; i < PS; i++) {
				indi trialindi = new indi(); //trial individual
				clone(trialindi, pop[i]);
				int rndD = (int) Math.floor(rnd.nextDouble()*D);
				int[] r = new int[3];
				rndInt(i, 3, r);
				for (int j = 0; j < D; j++) {					
					if (rnd.nextDouble()<=CR || rndD==j) {
						trialindi.x[j] = pop[r[0]].x[j] + F*(pop[r[1]].x[j]-pop[r[2]].x[j]); // rand/1/bin
						if (trialindi.x[j]<LOW[j] || UP[j]<trialindi.x[j]) 
							trialindi.x[j] = LOW[j] + rnd.nextDouble() * (UP[j] - LOW[j]);
					}
				} //end for D
				trialindi.fitness = Six_hump_Camel_back.f(trialindi.x);
				if (trialindi.fitness < pop[i].fitness)
					clone(pop[i], trialindi);
			} //end for PS
			increaseFEs(PS);
		} //end while
	}

	
	/*
	 * Get the index of the best individual
	 */
	public void getBest() {
		bestIdx = 0;
		for (int i = 1; i < PS; i++) {
			if (pop[i].fitness < pop[bestIdx].fitness)
				bestIdx = i;
		}
	}

}
